CONTRIBUTORS.md

```diff
@@ -0,0 +1,47 @@
+- akka extension (called `extension` from now on)  - this is where all custom metrics computation happens.
+- application that is being instrumented - in repo example, but this in theory this could be any application that uses akka.
+
+The most interesting parts are java agent and akka extension. They have to work in tandem - `agent` creates a foundation on which `extension` can build and calculate metrics.
+It's good to use an example to get a better understanding - `agent` allows to trigger an event whenever http request starts and completed 
+(we found places where those should be triggered by looking at akka source code and then modify those classes during load). 
+`extension` then catch those events and calculate duration between the two and save it in opentelemetry instrument. There will be more about opentelemetry role in the product later in this document.
+
+
+## Extension
+
+It's a part that is responsible for calculating and exporting metrics. The fact that it's akka extension is an implementation detail, but it's worth to know that.
+Akka extension are plugins that grant additional functionality to akka and can be enabled with configuration - this is the approach that we took.
+At the start up of the extension we check for configured metrics that were enabled and start only those monitors that are required.
+In our example monitors are akka system actors, but this is also an implementation detail.
+As an example `io.scalac.mesmer.extension.HttpEventsActor` is a monitor that listens for AkkaHttp events.
+For conveyor of messages we use Reception
+
+
+## Opentelemetry
+
+It's a fusion of opentracing and openCensus.  It's an specification and an implementation for creating, aggregating and publishing traces and metrics.
+We use this extensively as it's makes publishing metrics to different backends (prometheus, datadog etc) easy and can take care of some aggregations (mostly histograms ATM).
+Links to read more:
+- https://opentelemetry.io/docs/instrumentation/java/
+- https://github.com/open-telemetry/opentelemetry-specification
+- https://github.com/open-telemetry/opentelemetry-java
+
+
+## Agent
+
+This is probably the most complicated part of mesmer development  (there is an ongoing proposal to switch from our custom agent to opentelemetry one which would  simplify development).
+In some cases it requires writing `.java` files as scala doesn't allow some operations that are needed  for class instrumentations.
+We use bytebyddy as a preferred tool for bytecode generation - it operates on higher level that ASM and has an `Advice` abstraction - this can be thing as a blueprint that will be injected into loaded class file.
+Advice api is mostly annotation driven.
+
+Links to read more:
+- https://bytebuddy.net/#/tutorial
+- https://www.youtube.com/watch?v=oflzFGONG08 - introduction to bytebuddy and advices by it's creator Rafael Winterhalter
```

Comment:
Can you add a "how to add a new metric section"  below that will contain steps to do that? Even I could benefit from that (still not 100% sure what the steps are).

Reply:
Yeah, I must admit I didn't really follow a well thought process - but if I were to summarise is it would be a couple of simple steps:
* take a metric from the matrix
* check in akka source code how it can be implemented - this it the most tricky part, as we don't really know if something can be achieved and in which way. It's worth to know how akka works internally but this part is mostly trail and error. Sometimes, especially in "more modern" parts of akka code like `akka-persistence` or `akka-cluster` there is `@InternalStableApi` on methods that are stubs and probably overriden by lightbend monitoring agent.
* add exporting and if needed calculations in `extension`

Reply:
But I'll think if I can add something here as I see that this is not a huge help

Reply:
Something along these lines would be valuable indeed. For example: 

1. choose an API that you want to instrument with metrics
2. Make sure that you are able to test this instrumentation - if it's an Akka library, see if our example app uses that. If not, create a simple example app for testing that metric (consider contributing it to the "examples" directory in Mesmer.
3. Check how the instrumentation can be instrumented. What class should be modified by Bytebuddy? What method should report the metrics. What are the metrics, what is their kind (counter, gauge, histogram)?

I would also add something about: 

4. What is Agent* class? How to add it? How to enable it for the agent? (Show an example of an agent class)
5. What is event bus? What is it's purpose?
6. What is an <Metric>Actor class and what does it do? Where's the place where opentelemetry metrics API is actually used?

4, 5, 6 will possibly simplify in the future but for now it's relevant.